File: nodejs/src/skills/substitution.ts
/** Variable substitution for skill content. */

export interface SubstitutionContext {
  args?: string;
  skillDir?: string;
}

/**
 * Substitute variables in skill content:
 * - $ARGUMENTS       → user-provided arguments
 * - ${CLAUDE_SKILL_DIR} → skill directory path
 * - ${ENV:VAR_NAME}  → environment variable
 */
export function substituteVariables(content: string, ctx: SubstitutionContext): string {
  let result = content;

  if (ctx.args !== undefined) {
    result = result.replace(/\$ARGUMENTS/g, ctx.args);
  }

  if (ctx.skillDir !== undefined) {
    result = result.replace(/\$\{CLAUDE_SKILL_DIR\}/g, ctx.skillDir);
  }

  result = result.replace(/\$\{ENV:([^}]+)\}/g, (_, name: string) => {
    return process.env[name] || '';
  });

  return result;
}

File: go/skills/substitution.go
package skills

import (
	"os"
	"regexp"
)

// SubstituteVariables replaces variable placeholders in skill content.
//
// Supported variables:
//   - $ARGUMENTS           replaced with the user-provided arguments string
//   - ${CLAUDE_SKILL_DIR}  replaced with the skill's directory path
//   - ${ENV:VAR_NAME}      replaced with the value of the environment variable VAR_NAME
//
// This is a pure function with no side effects.
func SubstituteVariables(content string, args string, skillDir string) string {
	result := content

	if args != "" {
		re := regexp.MustCompile(`\$ARGUMENTS`)
		result = re.ReplaceAllString(result, args)
	}

	if skillDir != "" {
		re := regexp.MustCompile(`\$\{CLAUDE_SKILL_DIR\}`)
		result = re.ReplaceAllString(result, skillDir)
	}

	// Replace ${ENV:VAR_NAME} with the corresponding environment variable value.
	re := regexp.MustCompile(`\$\{ENV:([^}]+)\}`)
	result = re.ReplaceAllStringFunc(result, func(match string) string {
		submatch := re.FindStringSubmatch(match)
		if len(submatch) > 1 {
			val := os.Getenv(submatch[1])
			return val
		}
		return match
	})

	return result
}
